fasm/__init__.py

```diff
@@ -46,122 +60,134 @@ class ValueFormat(enum.Enum):
 # comment should a string or None.
 FasmLine = namedtuple('FasmLine', 'set_feature annotations comment')
 
-
-def assert_max_width(width, value):
-    """ asserts if the value is greater than the width. """
-    assert value < (2**width), (width, value)
-
-
-def verilog_value_to_int(verilog_value):
-    """ Convert VerilogValue model to width, value, value_format """
-    width = None
-
-    if verilog_value.plain_decimal:
-        return width, int(verilog_value.plain_decimal), ValueFormat.PLAIN
-
-    if verilog_value.width:
-        width = int(verilog_value.width)
-
-    if verilog_value.hex_value:
-        value = int(verilog_value.hex_value.replace('_', ''), 16)
-        value_format = ValueFormat.VERILOG_HEX
-    elif verilog_value.binary_value:
-        value = int(verilog_value.binary_value.replace('_', ''), 2)
-        value_format = ValueFormat.VERILOG_BINARY
-    elif verilog_value.decimal_value:
-        value = int(verilog_value.decimal_value.replace('_', ''), 10)
-        value_format = ValueFormat.VERILOG_DECIMAL
-    elif verilog_value.octal_value:
-        value = int(verilog_value.octal_value.replace('_', ''), 8)
-        value_format = ValueFormat.VERILOG_OCTAL
+# The following functions decode the binary format generated by parse_fasm.
+# This is a lightweight binary format designed to be fast to decode.
+
+# Match a tag and retrieve length from the header.
+def get_header(tags, data, i):
+    if not data[i] in bytes(tags, 'ascii'):
+        # print('tag mismatch: got {}, expecting {} at position {}'.format(data[i], tags, i))
+        # print('data: {}', data[:i])
+        return None, i
+
+    return int.from_bytes(data[i+1:i+5], byteorder), i+5
+
+# Decode a tagged string
+def tagged_string_from_bytes(tag, data, i):
+    length, i = get_header(tag, data, i)
+    if length is None:
+        return None, i
+
+    return data[i:i+length].decode('ascii'), i+length
+
+# Decode a FASM address: [x:y]
+def fasm_address_from_bytes(data, i):
+    length, i = get_header(':', data, i)
+    if not length:
+        return None, None, i
+
+    end = int.from_bytes(data[i:i+4], byteorder)
```

Comment:
`length` can only be 4 or 8 here, but you are not validating that?

Reply:
I've added an assertion.